text/0000-eye-of-sauron.md

```diff
@@ -0,0 +1,256 @@
+
+One common example where this is a nuisance is when using iterator adapters, working which can easily create multiple-reference types such as `&&u32`. For example:
+    
+```Rust
+    let v : Vec<u32> = vec![0, 1, 2, 3];
+    // Why can't I just write `x > 1`? Why do I have to ** for the compiler???
+    v.iter().filter(|x| **x > 1).count();
+```
+
+There are several cases where these operators are used. The most popular case is indexing associative maps:
+
+```Rust
+    use std::collections::HashMap;
+    let x: HashMap<_, _> =
+        vec![(format!("hello"), format!("world"))].into_iter().collect();
+    let s = format!("hello");
+    
+    // I would like to write...
+    println!("{}", x[s]);
+    // But instead, I have to write...
+    println!("{}", x[&s]);
+```
+
+In this case, these are merely annoying papercuts. In some other cases, they can be a much worse problem.
+
+One of thim is Isis Lovecruft's "Eye of Sauron" case, which is a problem when using non-`Copy` bignums:
+
+```Rust,ignore
+struct Bignum(..);
+// fair enough
+impl<'a> Add for &'a Bignum {
+    type Output = Bignum;
+    // ...
+}
+// ...
+
+// I can't see what my code is doing! It looks like one big piece of &.
+let a = &(-(&A)) * &(&one + &nrr).invert();
+let z = &u * &(&(&u.square() + &(&A * &u)) + &one);
+
+// Would be better:
+let a = (-A) * (one + nrr).invert();
+let z = u * (u.square() + A * u + one);
+```
+
+Allowing method-style autoref could make operator uses as clean as methods and fields, making code that uses them intensively far more ergonomic. It's also a fairly non-invasive extension.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Like methods, operators and indexing support automatic referencing and dereferencing. When you use an operator, the compiler will automatically add `&`, `&mut` and `*` operators to match the signature of the operator.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+"Operators" here refers to several typeck operators:
+- The "standard" eager binary operators: `+`, `-`, `*`, `/`, `%`, `^`, `&`, `|`, `<<`, `>>`, `==`, `<`, `<=`, `!=`, `>=`, `>`.
+- The "standard" unary operators: `!`, `-`. This does *NOT* include the dereference operator `*`.
+- The indexing operator `...[...]`
+
+Operator type-checking behaves similarly to method type-checking. It works as follows:
+
+## S1. Subexpression checking
+
+Both the LHS and the RHS (if binary) of the operator are first type-checked with no expected type.
+
+This differs from rustc 1.20, in which an expected type was sometimes propagated from the LHS into the RHS, potentially triggering coercions within the RHS. I should probably come up with an example in which this matters.
```

Comment:
This one's probably a crazy example, but the RHS is definitely dependent on the LHS and the case is rather degenerate:

```rust
    let byte = 0u8;
    let word = 1u16;

    let x = word + &byte.into();
    println!("{}", x);
```

This needs the `&` for *$reasons*

Found in https://github.com/rust-lang-nursery/rust-clippy/issues/2042

Reply:
> This one's probably a crazy example, but the RHS is definitely dependent on the LHS and the case is rather degenerate:

That's a trait-system issue. Operator overloading does succeed in this case. Without the autoref, you get these trait bounds:
```Rust
u16: Add<$0>
u8: Into<$0>
```

Because the compiler only considers each trait bound alone, it can't see that the only solution is `$0 = u16`.

If you add a reference, instead yoou get
```Rust
u16: Add<&'a $0>
u8: Into<$0>
```

And on the first bound, only the `impl Add<&u16> for u16` impl matches, which allows a solution.

Reply:
This also won't coerce: `&&1 + 2`. As soon as you have two different implementations for the RHS it stops working. That coercion only fires in such rare situations that I wonder if it's existence is by design or an accident of implementation.

Reply:
In any case it's orthogonal to this RFC.